# Problem
Sprocket Central Pty Ltd , a medium size bikes & cycling accessories organisation and would like to expand their product to the new market. Since Sprocket Central is keen to KPMG analytics, following analysis is to help Sprocket Central optimise its marketing strategy with provided dataset.


Dataset overview:

    - Worksheets 
        - Transaction data in the past 3 months
        - New Customer List
        - Customer Demographic
        - Customer Addresses
    - Collected in 2017

# Data preparation for analysis
Review data quality to ensure it is ready for further analysis and data issues are collected for client to mitigate current data quality. 
 

In [2]:
#import libraries
import pandas as pd
import numpy as np

In [3]:
df = pd.ExcelFile("SprocketCentral_raw_data.xlsx")
df.sheet_names

['Transactions', 'NewCustomerList', 'CustomerDemographic', 'CustomerAddress']

In [ ]:
#open separately worksheet 
Transactions = pd.read_excel(df, sheet_name="Transactions", header=1)
NewCustomerList = pd.read_excel(df, sheet_name="NewCustomerList", header=1)
CustomerDemographic = pd.read_excel(df, sheet_name="CustomerDemographic", header=1)
CustomerAddress = pd.read_excel(df, sheet_name="CustomerAddress", header=1)

/var/folders/yt/z2b0qpf52fxbhbtlsny218m00000gn/T/ipykernel_32691/1871028544.py:3: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype=datetime64[ns])
  NewCustomerList = pd.read_excel(df, sheet_name="NewCustomerList", header=1)
/var/folders/yt/z2b0qpf52fxbhbtlsny218m00000gn/T/ipykernel_32691/1871028544.py:4: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype=datetime64[ns])
  CustomerDemographic = pd.read_excel(df, sheet_name="CustomerDemographic", header=1)


## Transactions
In the Transactions data sheet, for customer insight analysis purpose, 3 critical information that must be sufficient are customer_id, product information (brand, product_line, product_class, product_size) and list_price. These information will be grouped to know which product is customer favorite.

In [ ]:
Transactions

transaction_id  product_id  customer_id transaction_date  online_order  \
0                   1           2         2950       2017-02-25           0.0   
1                   2           3         3120       2017-05-21           1.0   
2                   3          37          402       2017-10-16           0.0   
3                   4          88         3135       2017-08-31           0.0   
4                   5          78          787       2017-10-01           1.0   
...               ...         ...          ...              ...           ...   
19995           19996          51         1018       2017-06-24           1.0   
19996           19997          41          127       2017-11-09           1.0   
19997           19998          87         2284       2017-04-14           1.0   
19998           19999           6         2764       2017-07-03           0.0   
19999           20000          11         1144       2017-09-22           1.0   

      order_status           brand product_line product_class product_size  \
0         Approved           Solex     Standard        medium       medium   
1         Approved   Trek Bicycles     Standard        medium        large   
2         Approved      OHM Cycles     Standard           low       medium   
3         Approved  Norco Bicycles     Standard        medium       medium   
4         Approved  Giant Bicycles     Standard        medium        large   
...            ...             ...          ...           ...          ...   
19995     Approved      OHM Cycles     Standard          high       medium   
19996     Approved           Solex         Road        medium       medium   
19997     Approved      OHM Cycles     Standard        medium       medium   
19998     Approved      OHM Cycles     Standard          high       medium   
19999     Approved   Trek Bicycles     Standard        medium        small   

       list_price  standard_cost  product_first_sold_date  
0           71.49          53.62                  41245.0  
1         2091.47         388.92                  41701.0  
2         1793.43         248.82                  36361.0  
3         1198.46         381.10                  36145.0  
4         1765.30         709.48                  42226.0  
...           ...            ...                      ...  
19995     2005.66        1203.40                  37823.0  
19996      416.98         312.74                  35560.0  
19997     1636.90          44.71                  40410.0  
19998      227.88         136.73                  38216.0  
19999     1775.81        1580.47                  36334.0  

[20000 rows x 13 columns]

In [ ]:
#convert product_first_sold_date values format to readable datetime format
Transactions['product_first_sold_date'] = pd.to_datetime(Transactions['product_first_sold_date'], unit='s')
Transactions['product_first_sold_date']

0       1970-01-01 11:27:25
1       1970-01-01 11:35:01
2       1970-01-01 10:06:01
3       1970-01-01 10:02:25
4       1970-01-01 11:43:46
                ...        
19995   1970-01-01 10:30:23
19996   1970-01-01 09:52:40
19997   1970-01-01 11:13:30
19998   1970-01-01 10:36:56
19999   1970-01-01 10:05:34
Name: product_first_sold_date, Length: 20000, dtype: datetime64[ns]

product_first_sold_date column values are not correct as it shows everything happening the same day at different times. We would need to refer back to the author of the data to seek more classification on how the data was collected.

In [ ]:
#drop product_first_sold_date column
Transactions = Transactions.drop(['product_first_sold_date'], axis=1)

### a. Missing values
Missing values can be decided to drop depends on further analysis

In [ ]:
Transactions.isnull().sum()

transaction_id        0
product_id            0
customer_id           0
transaction_date      0
online_order        360
order_status          0
brand               197
product_line        197
product_class       197
product_size        197
list_price            0
standard_cost       197
dtype: int64

Since online_order values play a key role in grouping customer buying behavior, we need to drop 360 missing values in this case.  
 
Also with 197 missing values in brand values.

In [ ]:
#drop nan in online_order
Transactions = Transactions.dropna()

In [ ]:
Transactions.isnull().sum()

transaction_id      0
product_id          0
customer_id         0
transaction_date    0
online_order        0
order_status        0
brand               0
product_line        0
product_class       0
product_size        0
list_price          0
standard_cost       0
dtype: int64

In [ ]:
Transactions.head(10)

NameError: name 'Transactions' is not defined

### b. Duplicated values
Duplicated values can be decided to drop depends on further analysis

In [ ]:
#check duplicated value 
Transactions[Transactions.duplicated()].sum()

/var/folders/yt/z2b0qpf52fxbhbtlsny218m00000gn/T/ipykernel_32691/1752500654.py:2: FutureWarning: The default value of numeric_only in DataFrame.sum is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  Transactions[Transactions.duplicated()].sum()


transaction_id    0.0
product_id        0.0
customer_id       0.0
online_order      0.0
order_status      0.0
brand             0.0
product_line      0.0
product_class     0.0
product_size      0.0
list_price        0.0
standard_cost     0.0
dtype: float64

### c. Exploring data

In [ ]:
Transactions['order_status'].value_counts()

Approved     19273
Cancelled      172
Name: order_status, dtype: int64

In [ ]:
Transactions['brand'].value_counts()

Solex             4169
WeareA2B          3245
Giant Bicycles    3244
OHM Cycles        2993
Trek Bicycles     2931
Norco Bicycles    2863
Name: brand, dtype: int64

In [ ]:
Transactions['product_line'].value_counts()

Standard    13920
Road         3894
Touring      1213
Mountain      418
Name: product_line, dtype: int64

In [ ]:
Transactions['product_class'].value_counts()

medium    13587
high       2952
low        2906
Name: product_class, dtype: int64

In [ ]:
Transactions['product_size'].value_counts()

medium    12767
large      3900
small      2778
Name: product_size, dtype: int64

In [ ]:
pd.notna(Transactions['customer_id'].unique()).sum()

3492

In [ ]:
Transactions['customer_id'].value_counts()

1068    14
2476    14
2183    14
1302    13
2912    13
        ..
3392     1
2271     1
2328     1
1865     1
3161     1
Name: customer_id, Length: 3492, dtype: int64

Some customer_id bought more than 2 times, this information is for later analysis.

In [ ]:
print('customer_ids not in demographics dataset:', sum([(1 if (i not in CustomerDemographic['customer_id']) else 0) for i in CustomerAddress['customer_id']]))
print('customer_ids not in addresses dataset:', sum([(1 if (i not in CustomerDemographic['customer_id']) else 0) for i in Transactions['customer_id']]))

customer_ids not in demographics dataset: 4
customer_ids not in addresses dataset: 3


In [ ]:
#calculate profit
Transactions['profit'] = Transactions['list_price'] - Transactions['standard_cost']
print(Transactions[['list_price', 'standard_cost','profit']])

       list_price  standard_cost   profit
0           71.49          53.62    17.87
1         2091.47         388.92  1702.55
2         1793.43         248.82  1544.61
3         1198.46         381.10   817.36
4         1765.30         709.48  1055.82
...           ...            ...      ...
19995     2005.66        1203.40   802.26
19996      416.98         312.74   104.24
19997     1636.90          44.71  1592.19
19998      227.88         136.73    91.15
19999     1775.81        1580.47   195.34

[19445 rows x 3 columns]


In [ ]:
Transactions['profit'].describe()

count    19445.000000
mean       551.751933
std        493.211843
min          4.800000
25%        133.780000
50%        445.210000
75%        830.240000
max       1702.550000
Name: profit, dtype: float64

## New Customer List

In [ ]:
NewCustomerList

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0      Chickie    Brister    Male                                   86   
1        Morly     Genery    Male                                   69   
2      Ardelis  Forrester  Female                                   10   
3       Lucine      Stutt  Female                                   64   
4      Melinda     Hadlee  Female                                   34   
..         ...        ...     ...                                  ...   
995  Ferdinand  Romanetti    Male                                   60   
996       Burk    Wortley    Male                                   22   
997   Melloney      Temby  Female                                   17   
998     Dickie    Cubbini    Male                                   30   
999      Sylas    Duffill    Male                                   56   

           DOB                     job_title job_industry_category  \
0   1957-07-12               General Manager         Manufacturing   
1   1970-03-22           Structural Engineer              Property   
2   1974-08-28        Senior Cost Accountant    Financial Services   
3   1979-01-28    Account Representative III         Manufacturing   
4   1965-09-21             Financial Analyst    Financial Services   
..         ...                           ...                   ...   
995 1959-10-07                     Paralegal    Financial Services   
996 2001-10-17        Senior Sales Associate                Health   
997 1954-10-05  Budget/Accounting Analyst IV    Financial Services   
998 1952-12-17             Financial Advisor    Financial Services   
999 1955-10-02           Staff Accountant IV              Property   

        wealth_segment deceased_indicator owns_car  ...  state    country  \
0        Mass Customer                  N      Yes  ...    QLD  Australia   
1        Mass Customer                  N       No  ...    NSW  Australia   
2    Affluent Customer                  N       No  ...    VIC  Australia   
3    Affluent Customer                  N      Yes  ...    QLD  Australia   
4    Affluent Customer                  N       No  ...    NSW  Australia   
..                 ...                ...      ...  ...    ...        ...   
995  Affluent Customer                  N       No  ...    NSW  Australia   
996      Mass Customer                  N       No  ...    NSW  Australia   
997  Affluent Customer                  N      Yes  ...    QLD  Australia   
998      Mass Customer                  N      Yes  ...    QLD  Australia   
999      Mass Customer                  N      Yes  ...    NSW  Australia   

     property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                     6        0.76      0.9500     1.187500     1.009375   
1                    11        0.73      0.7300     0.912500     0.775625   
2                     5        1.02      1.0200     1.020000     1.020000   
3                     1        0.56      0.7000     0.700000     0.700000   
4                     9        0.68      0.6800     0.850000     0.850000   
..                  ...         ...         ...          ...          ...   
995                   7        0.74      0.7400     0.740000     0.740000   
996                  10        0.91      0.9100     1.137500     0.966875   
997                   2        0.96      1.2000     1.200000     1.200000   
998                   2        0.99      1.2375     1.237500     1.051875   
999                   9        0.95      1.1875     1.484375     1.261719   

     Unnamed: 20  Rank     Value  
0              1     1  1.718750  
1              1     1  1.718750  
2              1     1  1.718750  
3              4     4  1.703125  
4              4     4  1.703125  
..           ...   ...       ...  
995          996   996  0.374000  
996          997   997  0.357000  
997          997   997  0.357000  
998          997   997  0.357000  
999         1000  1000  0.340000  

[10

In [ ]:
NewCustomerList.columns

Index(['first_name', 'last_name', 'gender',
       'past_3_years_bike_related_purchases', 'DOB', 'job_title',
       'job_industry_category', 'wealth_segment', 'deceased_indicator',
       'owns_car', 'tenure', 'address', 'postcode', 'state', 'country',
       'property_valuation', 'Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18',
       'Unnamed: 19', 'Unnamed: 20', 'Rank', 'Value'],
      dtype='object')

In [ ]:
NewCustomerList = NewCustomerList.drop(['Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18',
       'Unnamed: 19', 'Unnamed: 20'], axis=1)

### a. Missing values
Missing values can be decided to drop depends on further analysis

In [ ]:
NewCustomerList.isnull().sum()

first_name                               0
last_name                               29
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     17
job_title                              106
job_industry_category                  165
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
address                                  0
postcode                                 0
state                                    0
country                                  0
property_valuation                       0
Rank                                     0
Value                                    0
dtype: int64

DOB, job_title values play a key role in grouping customer demographic, we need to drop missing values in this case.

In [ ]:
NewCustomerList = NewCustomerList.dropna()

In [ ]:
NewCustomerList.isnull().sum()

first_name                             0
last_name                              0
gender                                 0
past_3_years_bike_related_purchases    0
DOB                                    0
job_title                              0
job_industry_category                  0
wealth_segment                         0
deceased_indicator                     0
owns_car                               0
tenure                                 0
address                                0
postcode                               0
state                                  0
country                                0
property_valuation                     0
Rank                                   0
Value                                  0
dtype: int64

In [ ]:
NewCustomerList.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  tenure              address  \
0      Mass Customer                  N      Yes      14     45 Shopko Center   
1      Mass Customer                  N       No      16    14 Mccormick Park   
2  Affluent Customer                  N       No      10  5 Colorado Crossing   
3  Affluent Customer                  N      Yes       5   207 Annamark Plaza   
4  Affluent Customer                  N       No      19    115 Montana Place   

   postcode state    country  property_valuation  Rank     Value  
0      4500   QLD  Australia                   6     1  1.718750  
1      2113   NSW  Australia                  11     1  1.718750  
2      3505   VIC  Australia                   5     1  1.718750  
3      4814   QLD  Australia                   1     4  1.703125  
4      2093   NSW  Australia                   9     4  1.703125

### b. Duplicated values
Duplicated values can be decided to drop depends on further analysis.

In [ ]:
#check duplicated value 
NewCustomerList[NewCustomerList.duplicated()].sum()

/var/folders/yt/z2b0qpf52fxbhbtlsny218m00000gn/T/ipykernel_32691/309158485.py:2: FutureWarning: The default value of numeric_only in DataFrame.sum is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  NewCustomerList[NewCustomerList.duplicated()].sum()


first_name                             0.0
last_name                              0.0
gender                                 0.0
past_3_years_bike_related_purchases    0.0
job_title                              0.0
job_industry_category                  0.0
wealth_segment                         0.0
deceased_indicator                     0.0
owns_car                               0.0
tenure                                 0.0
address                                0.0
postcode                               0.0
state                                  0.0
country                                0.0
property_valuation                     0.0
Rank                                   0.0
Value                                  0.0
dtype: float64

### c. Exploring data

In [ ]:
pd.notna(NewCustomerList['first_name'].unique()).sum()

682

In [ ]:
NewCustomerList['DOB'].describe()

/var/folders/yt/z2b0qpf52fxbhbtlsny218m00000gn/T/ipykernel_32691/2634456645.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  NewCustomerList['DOB'].describe()


count                     715
unique                    702
top       1987-01-15 00:00:00
freq                        2
first     1938-06-08 00:00:00
last      2002-01-17 00:00:00
Name: DOB, dtype: object

In [ ]:
#add age column from DOB
now = pd.Timestamp('now')
NewCustomerList['Age'] = (now - NewCustomerList['DOB']).astype('<m8[Y]')
NewCustomerList

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0      Chickie    Brister    Male                                   86   
1        Morly     Genery    Male                                   69   
2      Ardelis  Forrester  Female                                   10   
3       Lucine      Stutt  Female                                   64   
4      Melinda     Hadlee  Female                                   34   
..         ...        ...     ...                                  ...   
995  Ferdinand  Romanetti    Male                                   60   
996       Burk    Wortley    Male                                   22   
997   Melloney      Temby  Female                                   17   
998     Dickie    Cubbini    Male                                   30   
999      Sylas    Duffill    Male                                   56   

           DOB                     job_title job_industry_category  \
0   1957-07-12               General Manager         Manufacturing   
1   1970-03-22           Structural Engineer              Property   
2   1974-08-28        Senior Cost Accountant    Financial Services   
3   1979-01-28    Account Representative III         Manufacturing   
4   1965-09-21             Financial Analyst    Financial Services   
..         ...                           ...                   ...   
995 1959-10-07                     Paralegal    Financial Services   
996 2001-10-17        Senior Sales Associate                Health   
997 1954-10-05  Budget/Accounting Analyst IV    Financial Services   
998 1952-12-17             Financial Advisor    Financial Services   
999 1955-10-02           Staff Accountant IV              Property   

        wealth_segment deceased_indicator owns_car  tenure  \
0        Mass Customer                  N      Yes      14   
1        Mass Customer                  N       No      16   
2    Affluent Customer                  N       No      10   
3    Affluent Customer                  N      Yes       5   
4    Affluent Customer                  N       No      19   
..                 ...                ...      ...     ...   
995  Affluent Customer                  N       No       9   
996      Mass Customer                  N       No       6   
997  Affluent Customer                  N      Yes      15   
998      Mass Customer                  N      Yes      19   
999      Mass Customer                  N      Yes      14   

                      address  postcode state    country  property_valuation  \
0            45 Shopko Center      4500   QLD  Australia                   6   
1           14 Mccormick Park      2113   NSW  Australia                  11   
2         5 Colorado Crossing      3505   VIC  Australia                   5   
3          207 Annamark Plaza      4814   QLD  Australia                   1   
4           115 Montana Place      2093   NSW  Australia                   9   
..                        ...       ...   ...        ...                 ...   
995               2 Sloan Way      2200   NSW  Australia                   7   
996         04 Union Crossing      2196   NSW  Australia                  10   
997  33475 Fair Oaks Junction      4702   QLD  Australia                   2   
998        57666 Victoria Way      4215   QLD  Australia                   2   
999        21875 Grover Drive      2010   NSW  Australia                   9   

     Rank     Value   Age  
0       1  1.718750  65.0  
1       1  1.718750  53.0  
2       1  1.718750  48.0  
3       4  1.703125  44.0  
4       4  1.703125  57.0  
..    ...       ...   ...  
995   996  0.374000  63.0  
996   997  0.357000  21.0  
997   997  0.357000  68.0  
998   997  0.357000  70.0  
999  1000  0.340000  67.0  

[715 rows x 19 columns]

In [ ]:
NewCustomerList['gender'].value_counts()

Female    369
Male      346
Name: gender, dtype: int64

In [ ]:
NewCustomerList['job_title'].value_counts()

Environmental Tech             12
Software Consultant            12
Associate Professor            12
Cost Accountant                12
Junior Executive               11
                               ..
Web Developer II                1
Programmer Analyst IV           1
Media Manager I                 1
Computer Systems Analyst IV     1
Quality Control Specialist      1
Name: job_title, Length: 176, dtype: int64

In [ ]:
NewCustomerList['owns_car'].value_counts()

No     366
Yes    349
Name: owns_car, dtype: int64

In [ ]:
NewCustomerList['state'].value_counts()

NSW    352
VIC    197
QLD    166
Name: state, dtype: int64

In [ ]:
NewCustomerList['Rank'].value_counts()

259     10
760      9
536      8
700      7
444      6
        ..
326      1
358      1
724      1
722      1
1000     1
Name: Rank, Length: 289, dtype: int64

## Customer Demographic 

In [ ]:
CustomerDemographic

customer_id      first_name  last_name  gender  \
0               1         Laraine  Medendorp       F   
1               2             Eli    Bockman    Male   
2               3           Arlin     Dearle    Male   
3               4          Talbot        NaN    Male   
4               5  Sheila-kathryn     Calton  Female   
...           ...             ...        ...     ...   
3995         3996         Rosalia   Halgarth  Female   
3996         3997          Blanch     Nisuis  Female   
3997         3998          Sarene    Woolley       U   
3998         3999       Patrizius        NaN    Male   
3999         4000           Kippy    Oldland    Male   

      past_3_years_bike_related_purchases        DOB               job_title  \
0                                      93 1953-10-12     Executive Secretary   
1                                      81 1980-12-16  Administrative Officer   
2                                      61 1954-01-20      Recruiting Manager   
3                                      33 1961-10-03                     NaN   
4                                      56 1977-05-13           Senior Editor   
...                                   ...        ...                     ...   
3995                                    8 1975-08-09   VP Product Management   
3996                                   87 2001-07-13         Statistician II   
3997                                   60        NaT       Assistant Manager   
3998                                   11 1973-10-24                     NaN   
3999                                   76 1991-11-05    Software Engineer IV   

     job_industry_category     wealth_segment deceased_indicator  \
0                   Health      Mass Customer                  N   
1       Financial Services      Mass Customer                  N   
2                 Property      Mass Customer                  N   
3                       IT      Mass Customer                  N   
4                      NaN  Affluent Customer                  N   
...                    ...                ...                ...   
3995                Health      Mass Customer                  N   
3996         Manufacturing     High Net Worth                  N   
3997                    IT     High Net Worth                  N   
3998         Manufacturing  Affluent Customer                  N   
3999                   NaN  Affluent Customer                  N   

                                                default owns_car  tenure  
0                                                    "'      Yes    11.0  
1                          <script>alert('hi')</script>      Yes    16.0  
2                                   2018-02-01 00:00:00      Yes    15.0  
3     () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                   NIL      Yes     8.0  
...                                                 ...      ...     ...  
3995                                               -100       No    19.0  
3996                                           â¦test⧠     Yes     1.0  
3997                                                NaN       No     NaN  
3998                             ¡â¢Â£Â¢â§¶â¢ÂªÂºââ       Yes    10.0  
3999                                                0/0       No    11.0  

[4000 rows x 13 columns]

### a. Invalid values

In [ ]:
CustomerDemographic['default'].value_counts()

100                                       113
1                                         112
-1                                        111
-100                                       99
١٢٣                                     53
                                         ... 
testâ testâ«                               31
/dev/null; touch /tmp/blns.fail ; echo     30
âªâªtest⪠                                29
ì¸ëë°í ë¥´                                 27
,ãã»:*:ã»ãâ( â» Ï â» )ãã»:*:ã»ãâ           25
Name: default, Length: 90, dtype: int64

Since value in default column is invalid and not able to recognize, we then drop this column.

In [ ]:
CustomerDemographic = CustomerDemographic.drop(['default'], axis=1)

In [ ]:
CustomerDemographic.dtypes

customer_id                                     int64
first_name                                     object
last_name                                      object
gender                                         object
past_3_years_bike_related_purchases             int64
DOB                                    datetime64[ns]
job_title                                      object
job_industry_category                          object
wealth_segment                                 object
deceased_indicator                             object
owns_car                                       object
tenure                                        float64
dtype: object

In [ ]:
CustomerDemographic['DOB'].describe()

/var/folders/yt/z2b0qpf52fxbhbtlsny218m00000gn/T/ipykernel_32691/2638702152.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  CustomerDemographic['DOB'].describe()


count                    3913
unique                   3448
top       1978-01-30 00:00:00
freq                        7
first     1843-12-21 00:00:00
last      2002-03-11 00:00:00
Name: DOB, dtype: object

Here the smallest DOB value is 1843-12-21 which 180 years old so it does not valid. Now we have to review other column value of this person to consider whether drop this data or adjust it.

In [ ]:
CustomerDemographic['DOB'] = pd.to_datetime(CustomerDemographic['DOB'])
CustomerDemographic = CustomerDemographic.sort_values('DOB')
CustomerDemographic.head(2)

customer_id first_name last_name gender  \
33            34   Jephthah  Bachmann      U   
719          720     Darrel     Canet   Male   

     past_3_years_bike_related_purchases        DOB           job_title  \
33                                    59 1843-12-21     Legal Assistant   
719                                   67 1931-10-23  Recruiting Manager   

    job_industry_category     wealth_segment deceased_indicator owns_car  \
33                     IT  Affluent Customer                  N       No   
719                Retail  Affluent Customer                  N       No   

     tenure  
33     20.0  
719     6.0

As customer_id number 34 (who mistyped DOB is 1843) consists of high number of past_3_years_bike_related_purchases, other column values are sufficient as well, we keep this data and adjust 1843 to 1943.

In [ ]:
CustomerDemographic.loc[CustomerDemographic['DOB'] == '1843-12-21', 'DOB'] = '1943-12-21'
CustomerDemographic.head(2)


customer_id first_name last_name gender  \
33            34   Jephthah  Bachmann      U   
719          720     Darrel     Canet   Male   

     past_3_years_bike_related_purchases        DOB           job_title  \
33                                    59 1943-12-21     Legal Assistant   
719                                   67 1931-10-23  Recruiting Manager   

    job_industry_category     wealth_segment deceased_indicator owns_car  \
33                     IT  Affluent Customer                  N       No   
719                Retail  Affluent Customer                  N       No   

     tenure  
33     20.0  
719     6.0

### b. Missing values
Missing values can be decided to drop depends on further analysis

In [ ]:
CustomerDemographic.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                  87
dtype: int64

In [ ]:
CustomerDemographic = CustomerDemographic.dropna()

### c. Inconsistency input values

In [ ]:
CustomerDemographic['gender'].value_counts()

Female    1444
Male      1333
U            1
F            1
Femal        1
Name: gender, dtype: int64

In [ ]:
CustomerDemographic['gender'] = CustomerDemographic['gender'].replace('F','Female').replace('Femal','Female').replace('M','Male')

### d. Exploring data

In [ ]:
#add age column from DOB
CustomerDemographic['Age'] = (now - CustomerDemographic['DOB']).astype('<m8[Y]')
CustomerDemographic

customer_id first_name    last_name  gender  \
33             34   Jephthah     Bachmann       U   
719           720     Darrel        Canet    Male   
1091         1092     Katlin      Creddon  Female   
2412         2413      Abbey       Murrow    Male   
657           658       Donn      Bonnell    Male   
...           ...        ...          ...     ...   
469           470     Darbee    Klimowicz    Male   
3948         3949      Costa  Sleightholm    Male   
2295         2296   Nathalia       Sanger  Female   
1887         1888      Sibyl      Scholtz  Female   
65             66     Anselm        Gawne    Male   

      past_3_years_bike_related_purchases        DOB  \
33                                     59 1943-12-21   
719                                    67 1931-10-23   
1091                                   56 1935-08-22   
2412                                   27 1943-08-11   
657                                    38 1944-01-24   
...                                   ...        ...   
469                                    22 2001-12-19   
3948                                   24 2001-12-19   
2295                                   16 2002-01-01   
1887                                   67 2002-01-26   
65                                     46 2002-03-11   

                     job_title job_industry_category     wealth_segment  \
33             Legal Assistant                    IT  Affluent Customer   
719         Recruiting Manager                Retail  Affluent Customer   
1091        VP Quality Control                Retail      Mass Customer   
2412  Environmental Specialist         Manufacturing     High Net Worth   
657             Tax Accountant         Manufacturing  Affluent Customer   
...                        ...                   ...                ...   
469           Internal Auditor                Retail      Mass Customer   
3948           Web Designer IV         Manufacturing     High Net Worth   
2295              Geologist IV                Retail  Affluent Customer   
1887              Food Chemist                Health      Mass Customer   
65           Account Executive           Argiculture     High Net Worth   

     deceased_indicator owns_car  tenure   Age  
33                    N       No    20.0  79.0  
719                   N       No     6.0  91.0  
1091                  N       No     5.0  87.0  
2412                  N      Yes    17.0  79.0  
657                   N      Yes     8.0  79.0  
...                 ...      ...     ...   ...  
469                   N       No     1.0  21.0  
3948                  N      Yes     1.0  21.0  
2295                  N       No     1.0  21.0  
1887                  N      Yes     1.0  21.0  
65                    N       No     1.0  21.0  

[2780 rows x 13 columns]

In [ ]:
#check unique customer_id
pd.notna(CustomerDemographic['customer_id'].unique()).sum()

2780

In [ ]:
CustomerDemographic['gender'].value_counts()

Female    1446
Male      1333
U            1
Name: gender, dtype: int64

## Customer Addresses

In [ ]:
CustomerAddress

customer_id                    address  postcode            state  \
0               1         060 Morning Avenue      2016  New South Wales   
1               2        6 Meadow Vale Court      2153  New South Wales   
2               4         0 Holy Cross Court      4211              QLD   
3               5        17979 Del Mar Point      2448  New South Wales   
4               6           9 Oakridge Court      3216              VIC   
...           ...                        ...       ...              ...   
3994         3999            1482 Hauk Trail      3064              VIC   
3995         4000  57042 Village Green Point      4511              QLD   
3996         4001     87 Crescent Oaks Alley      2756              NSW   
3997         4002           8194 Lien Street      4032              QLD   
3998         4003            320 Acker Drive      2251              NSW   

        country  property_valuation  
0     Australia                  10  
1     Australia                  10  
2     Australia                   9  
3     Australia                   4  
4     Australia                   9  
...         ...                 ...  
3994  Australia                   3  
3995  Australia                   6  
3996  Australia                  10  
3997  Australia                   7  
3998  Australia                   7  

[3999 rows x 6 columns]

Since Customer Demographic look up customer_id from Customer Addresses, there is 1 missing values in Customer Addresses (3999 rows) when Customer Demographic is 4000 rows.

### a. Missing values

In [ ]:
CustomerAddress.isnull().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

### b. Inconsistency input values

In [ ]:
CustomerAddress['state'].value_counts()


NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: state, dtype: int64

In [ ]:
CustomerAddress['state'] = CustomerAddress['state'].replace('New South Wales','NSW').replace('Victoria','VIC')

In [ ]:
CustomerAddress.head()

customer_id              address  postcode state    country  \
0            1   060 Morning Avenue      2016   NSW  Australia   
1            2  6 Meadow Vale Court      2153   NSW  Australia   
2            4   0 Holy Cross Court      4211   QLD  Australia   
3            5  17979 Del Mar Point      2448   NSW  Australia   
4            6     9 Oakridge Court      3216   VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

### c. Unique value

In [ ]:
CustomerAddress.nunique()

customer_id           3999
address               3996
postcode               873
state                    3
country                  1
property_valuation      12
dtype: int64

### d. Exploring data

In [ ]:
CustomerAddress['postcode'].value_counts()

2170    31
2155    30
2145    30
2153    29
3977    26
        ..
3808     1
3114     1
4721     1
4799     1
3089     1
Name: postcode, Length: 873, dtype: int64

In [ ]:
CustomerAddress['state'].value_counts()

NSW    2140
VIC    1021
QLD     838
Name: state, dtype: int64

## Merge Customer Demographic and Customer Addressess to Transactions 

In [ ]:
#get columns in CustomerDemographic that are not in Transactions
difference1 = CustomerDemographic.columns.difference(Transactions.columns)
print (difference1)

Index(['Age', 'DOB', 'deceased_indicator', 'first_name', 'gender',
       'job_industry_category', 'job_title', 'last_name', 'owns_car',
       'past_3_years_bike_related_purchases', 'tenure', 'wealth_segment'],
      dtype='object')


In [ ]:
merged_info1 = pd.merge(Transactions,CustomerDemographic[['customer_id','Age', 'gender', 'job_industry_category', 'job_title','owns_car','past_3_years_bike_related_purchases','wealth_segment' ]],on='customer_id', how='left')
merged_info1.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost   profit   Age  gender job_industry_category  \
0       71.49          53.62    17.87  68.0    Male    Financial Services   
1     2091.47         388.92  1702.55  44.0  Female                Health   
2     1793.43         248.82  1544.61  45.0    Male                Retail   
3     1198.46         381.10   817.36  61.0    Male    Financial Services   
4     1765.30         709.48  1055.82   NaN     NaN                   NaN   

                    job_title owns_car  past_3_years_bike_related_purchases  \
0         Software Engineer I      Yes                                 19.0   
1         Clinical Specialist      Yes                                 89.0   
2  Desktop Support Technician       No                                  9.0   
3             Staff Scientist       No                                 83.0   
4                         NaN      NaN                                  NaN   

      wealth_segment  
0      Mass Customer  
1      Mass Customer  
2  Affluent Customer  
3      Mass Customer  
4                NaN

In [ ]:
#get columns in CustomerAddresses that are not in CustomerDemographic
difference2 = CustomerAddress.columns.difference(CustomerDemographic.columns)
print (difference2)

Index(['address', 'country', 'postcode', 'property_valuation', 'state'], dtype='object')


In [ ]:
merged_info2 = pd.merge(merged_info1,CustomerAddress[['customer_id', 'postcode','state' ]],on='customer_id', how='left')
merged_info2.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  ...  \
0     Approved           Solex     Standard        medium       medium  ...   
1     Approved   Trek Bicycles     Standard        medium        large  ...   
2     Approved      OHM Cycles     Standard           low       medium  ...   
3     Approved  Norco Bicycles     Standard        medium       medium  ...   
4     Approved  Giant Bicycles     Standard        medium        large  ...   

    profit   Age  gender  job_industry_category                   job_title  \
0    17.87  68.0    Male     Financial Services         Software Engineer I   
1  1702.55  44.0  Female                 Health         Clinical Specialist   
2  1544.61  45.0    Male                 Retail  Desktop Support Technician   
3   817.36  61.0    Male     Financial Services             Staff Scientist   
4  1055.82   NaN     NaN                    NaN                         NaN   

  owns_car past_3_years_bike_related_purchases     wealth_segment  postcode  \
0      Yes                                19.0      Mass Customer    3064.0   
1      Yes                                89.0      Mass Customer    2196.0   
2       No                                 9.0  Affluent Customer    2835.0   
3       No                                83.0      Mass Customer    2096.0   
4      NaN                                 NaN                NaN    2292.0   

  state  
0   VIC  
1   NSW  
2   NSW  
3   NSW  
4   NSW  

[5 rows x 22 columns]

In [ ]:
merged_info2.isnull().sum()

transaction_id                            0
product_id                                0
customer_id                               0
transaction_date                          0
online_order                              0
order_status                              0
brand                                     0
product_line                              0
product_class                             0
product_size                              0
list_price                                0
standard_cost                             0
profit                                    0
Age                                    5797
gender                                 5797
job_industry_category                  5797
job_title                              5797
owns_car                               5797
past_3_years_bike_related_purchases    5797
wealth_segment                         5797
postcode                                 29
state                                    29
dtype: int64

In [ ]:
new_customer_data=merged_info2.dropna()
new_customer_data.isnull().sum()

transaction_id                         0
product_id                             0
customer_id                            0
transaction_date                       0
online_order                           0
order_status                           0
brand                                  0
product_line                           0
product_class                          0
product_size                           0
list_price                             0
standard_cost                          0
profit                                 0
Age                                    0
gender                                 0
job_industry_category                  0
job_title                              0
owns_car                               0
past_3_years_bike_related_purchases    0
wealth_segment                         0
postcode                               0
state                                  0
dtype: int64

In [ ]:
new_customer_data.columns

Index(['transaction_id', 'product_id', 'customer_id', 'transaction_date',
       'online_order', 'order_status', 'brand', 'product_line',
       'product_class', 'product_size', 'list_price', 'standard_cost',
       'profit', 'Age', 'gender', 'job_industry_category', 'job_title',
       'owns_car', 'past_3_years_bike_related_purchases', 'wealth_segment',
       'postcode', 'state'],
      dtype='object')

In [ ]:
#export cleaned data to an excel file for further analysis
#new_customer_data.to_excel("new_customer_data.xlsx")